File: Solution/8.string-to-integer-atoi.go
package LeetCode

import (
	"math"
	"unicode"
)

/*
 * @lc app=leetcode id=8 lang=golang
 *
 * [8] String to Integer (atoi)
 */

// @lc code=start

func myAtoi(s string) int {
	var (
		result     int // initialize result as 0
		isnegative bool
		started    bool
	)

	for _, v := range s {
		// check if the character is a digit
		if unicode.IsDigit(v) {
			started = true
			digit := int(v - '0')      // convert the character to a digit
			result = result*10 + digit // add the digit to the result
			// check if the result is out of range
			if !isnegative && result > math.MaxInt32 {
				return math.MaxInt32
			}
			if isnegative && -result < math.MinInt32 {
				return math.MinInt32
			}
		} else if v == '+' && !started {
			started = true
		} else if v == '-' && !started {
			started = true
			isnegative = true
		} else if v != ' ' || started {
			// exit the loop if non-space character is encountered after the
			// start
			break
		}
	}

	if isnegative {
		result *= -1
	}

	return result
}

// @lc code=end

File: 8.string-to-integer-atoi.py
#
# @lc app=leetcode id=8 lang=python
#
# [8] String to Integer (atoi)
#

# @lc code=start
import math


class Solution(object):
    def myAtoi(self, s):
        """
        :type s: str
        :rtype: int
        """
        result = 0
        isnegative = False
        started = False

        for v in s:
            if v.isdigit():
                started = True
                digit = int(v)
                result = result * 10 + digit
                if not isnegative and result > math.pow(2, 31) - 1:
                    return int(math.pow(2, 31) - 1)
                if isnegative and -result < -math.pow(2, 31):
                    return int(-math.pow(2, 31))
            elif v == "+" and not started:
                started = True
            elif v == "-" and not started:
                started = True
                isnegative = True
            elif v != " " or started:
                break

        if isnegative:
            result *= -1

        return result


# @lc code=end
